File: mns-reactor/src/main/java/cz/fb/manaus/reactor/betting/listener/ProbabilityComparator.java
package cz.fb.manaus.reactor.betting.listener;

import com.google.common.collect.Ordering;
import com.google.common.primitives.Doubles;
import cz.fb.manaus.core.model.RunnerPrices;
import cz.fb.manaus.core.model.Side;

import java.util.Comparator;
import java.util.Map;

public class ProbabilityComparator implements Comparator<RunnerPrices> {

    public static Map<Side, Ordering<RunnerPrices>> COMPARATORS = Map.of(
            Side.BACK, Ordering.from(new ProbabilityComparator(Side.BACK)),
            Side.LAY, Ordering.from(new ProbabilityComparator(Side.LAY)));
    private final Side side;

    private ProbabilityComparator(Side side) {
        this.side = side;
    }

    @Override
    public int compare(RunnerPrices list1, RunnerPrices list2) {
        var backList1 = list1.getHomogeneous(side);
        var backList2 = list2.getHomogeneous(side);
        var price1 = backList1.getBestPrice().get().getPrice();
        var price2 = backList2.getBestPrice().get().getPrice();
        return Doubles.compare(price1, price2);
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/betting/listener/ProbabilityComparator.kt
package cz.fb.manaus.reactor.betting.listener

import com.google.common.primitives.Doubles
import cz.fb.manaus.core.model.RunnerPrices
import cz.fb.manaus.core.model.Side

class ProbabilityComparator(private val side: Side) : Comparator<RunnerPrices> {

    override fun compare(list1: RunnerPrices, list2: RunnerPrices): Int {
        val backList1 = list1.by(side)
        val backList2 = list2.by(side)
        val bestPrice1 = backList1.bestPrice ?: error("empty price list")
        val bestPrice2 = backList2.bestPrice ?: error("empty price list")
        val price1 = bestPrice1.price
        val price2 = bestPrice2.price
        return Doubles.compare(price1, price2)
    }

    companion object {
        var COMPARATORS = mapOf(
                Side.BACK to ProbabilityComparator(Side.BACK),
                Side.LAY to ProbabilityComparator(Side.LAY)
        )
    }

}

fun sortPrices(side: Side, marketPrices: List<RunnerPrices>): List<RunnerPrices> {
    val ordering = ProbabilityComparator.COMPARATORS[side] ?: error("no such side")
    val sortedPrices = marketPrices.sortedWith(ordering)
    check(
            sortedPrices.map { it.selectionId }.distinct().count() ==
                    sortedPrices.map { it.selectionId }.count()
    )
    return sortedPrices
}
